lib/http1.c

```diff
@@ -783,6 +794,128 @@ static void on_upgrade_complete(h2o_socket_t *socket, const char *err)
     cb(data, sock, headers_size);
 }
 
+static void tunnel_close(struct st_h2o_http1_tunnel_t *tunnel, const char *err)
+{
+    /* TODO log err */
+    tunnel->server->destroy(tunnel->server);
+    if (tunnel->client != NULL)
+        h2o_socket_close(tunnel->client);
+    h2o_timer_unlink(&tunnel->idle_timeout.entry);
+
+    free(tunnel);
+}
+
+static void tunnel_on_timeout(h2o_timer_t *timer)
+{
+    struct st_h2o_http1_tunnel_t *tunnel = H2O_STRUCT_FROM_MEMBER(struct st_h2o_http1_tunnel_t, idle_timeout.entry, timer);
+
+    tunnel_close(tunnel, "idle timeout");
+}
+
+static void tunnel_reset_timeout(struct st_h2o_http1_tunnel_t *tunnel)
+{
+    h2o_timer_unlink(&tunnel->idle_timeout.entry);
+    h2o_timer_link(h2o_socket_get_loop(tunnel->client), tunnel->idle_timeout.ticks, &tunnel->idle_timeout.entry);
+}
+
+static void tunnel_on_client_write_complete(h2o_socket_t *_sock, const char *err)
+{
+    struct st_h2o_http1_tunnel_t *tunnel = _sock->data;
+    assert(tunnel->client == _sock);
+
+    if (err != NULL) {
+        tunnel_close(tunnel, err);
+        return;
+    }
+
+    tunnel->server->proceed_read(tunnel->server);
+}
+
+static void tunnel_on_server_read(h2o_httpclient_tunnel_t *_tunnel, const char *err, const void *bytes, size_t len)
+{
+    struct st_h2o_http1_tunnel_t *tunnel = _tunnel->data;
+    assert(tunnel->server == _tunnel);
+
+    if (err != NULL) {
+        tunnel_close(tunnel, err);
+        return;
+    }
+
+    tunnel_reset_timeout(tunnel);
+
+    h2o_iovec_t vec = h2o_iovec_init(bytes, len);
+    h2o_socket_write(tunnel->client, &vec, 1, tunnel_on_client_write_complete);
+}
+
+static void tunnel_on_client_read(h2o_socket_t *_sock, const char *err)
+{
+    struct st_h2o_http1_tunnel_t *tunnel = _sock->data;
+    assert(tunnel->client == _sock);
+
+    if (err != NULL) {
+        tunnel_close(tunnel, err);
+        return;
+    }
+
+    tunnel_reset_timeout(tunnel);
+
+    if (tunnel->client->input->size != 0) {
+        h2o_socket_read_stop(tunnel->client);
+        tunnel->server->write_(tunnel->server, tunnel->client->input->bytes, tunnel->client->input->size);
+    } else {
+        h2o_socket_read_start(tunnel->client, tunnel_on_client_read);
+    }
```

Comment:
We should have a comment stating that the socket never gets read between when `tunnel->write_` is called and when the `white_complete` callback is called. That is why, unlike many other parts of h2o, we can pass the content of `h2o_buffer_t *` (e.g., `client->input->bytes`) directly without fear of it being moved elsewhere due to buffer expansion.